Mathlib/Order/IsWellOrderLimitElement.lean

```diff
@@ -0,0 +1,104 @@
+/-
+Copyright (c) 2024 Joël Riou. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Joël Riou
+-/
+import Mathlib.Data.Nat.Order.Basic
+import Mathlib.Order.WellFounded
+
+/-!
+# Limit elements in well-ordered types
+
+This file introduces two main definitions:
+- `wellOrderSucc a`: the successor of an element `a` in a well-ordered type
+- the typeclass `IsWellOrderLimitElement a` which asserts that an element `a` (in a
+well-ordered type) is neither a successor nor the smallest element, i.e. `a` is a limit element
+
+Then, the lemma `eq_bot_or_eq_succ_or_isWellOrderLimitElement` shows that an element
+in a well-ordered type is either `⊥`, a successor, or a limit element.
+
+-/
+
+variable {α : Type*} [LinearOrder α] [IsWellOrder α (· < ·)]
+
+/-- Given an element `a : α` in a well ordered set, this is the successor of `a`,
+i.e. the smallest element stricly greater than `a` if it exists (or `a` itself otherwise). -/
+noncomputable def wellOrderSucc (a : α) : α :=
+  (@IsWellFounded.wf α (· < ·)).succ a
+
+lemma self_le_wellOrderSucc (a : α) : a ≤ wellOrderSucc a := by
+  by_cases h : ∃ b, a < b
+  · exact (IsWellFounded.wf.lt_succ h).le
+  · dsimp [wellOrderSucc, WellFounded.succ]
+    rw [dif_neg h]
+
+lemma wellOrderSucc_le {a b : α} (ha : a < b) : wellOrderSucc a ≤ b := by
+  dsimp [wellOrderSucc, WellFounded.succ]
+  rw [dif_pos ⟨_, ha⟩]
+  exact WellFounded.min_le _ ha
+
+lemma self_lt_wellOrderSucc {a b : α} (h : a < b) : a < wellOrderSucc a := by
+  dsimp [wellOrderSucc, WellFounded.succ]
+  rw [dif_pos ⟨b, h⟩]
+  apply WellFounded.min_mem
```

Comment:
```suggestion
lemma self_lt_wellOrderSucc {a b : α} (h : a < b) : a < wellOrderSucc a :=
  IsWellFounded.wf.lt_succ ⟨b, h⟩
```
I learnt the existence of this lemma from the lemma you proved ten lines earlier :-)

Reply:
Thanks!